pkg/storage/tsdb/bucketindex/loader.go

```diff
@@ -0,0 +1,253 @@
+	// and we overwrite it: last will win.
+	l.indexes[userID] = newCachedIndex(idx, err)
+}
+
+// updateCachedIndexes checks all cached indexes and, for each of them, does two things:
+// 1. Offload indexes not requested since >= idle timeout
+// 2. Update indexes which have been updated last time since >= update timeout
+func (l *Loader) updateCachedIndexes(ctx context.Context) error {
+	var (
+		toUpdate []string
+		toDelete []string
+	)
+
+	// Build a list of users for which we should update or delete the index.
+	l.indexesMx.RLock()
+	for userID, entry := range l.indexes {
+		if time.Since(entry.getRequestedAt()) >= l.cfg.IdleTimeout {
+			toDelete = append(toDelete, userID)
+		} else if entry.err != nil && time.Since(entry.getUpdatedAt()) >= l.cfg.UpdateOnErrorInterval {
+			toUpdate = append(toUpdate, userID)
+		} else if entry.err == nil && time.Since(entry.getUpdatedAt()) >= l.cfg.UpdateOnStaleInterval {
+			toUpdate = append(toUpdate, userID)
+		}
+	}
+	l.indexesMx.RUnlock()
+
+	// Delete unused indexes, if confirmed they're still unused.
+	l.indexesMx.Lock()
+	for _, userID := range toDelete {
+		if idx := l.indexes[userID]; time.Since(idx.getRequestedAt()) >= l.cfg.IdleTimeout {
+			delete(l.indexes, userID)
+			level.Info(l.logger).Log("msg", "unloaded bucket index", "user", userID, "reason", "idle")
+		}
+	}
+	l.indexesMx.Unlock()
+
+	// Update actively used indexes.
+	for _, userID := range toUpdate {
+		readCtx, cancel := context.WithTimeout(ctx, readIndexTimeout)
+		defer cancel()
+
+		l.loadAttempts.Inc()
+		startTime := time.Now()
+		idx, err := ReadIndex(readCtx, l.bkt, userID, l.logger)
+
+		if errors.Is(err, ErrIndexNotFound) {
+			level.Info(l.logger).Log("msg", "unloaded bucket index", "user", userID, "reason", "not found during periodic check")
+
+			// Remove from cache.
+			l.indexesMx.Lock()
+			delete(l.indexes, userID)
+			l.indexesMx.Unlock()
+
+			continue
+		}
+		if err != nil {
+			l.loadFailures.Inc()
+			level.Warn(l.logger).Log("msg", "unable to update bucket index", "user", userID, "err", err)
+			continue
+		}
+
+		l.loadDuration.Observe(time.Since(startTime).Seconds())
+
+		// Cache it.
+		l.indexesMx.Lock()
+		l.indexes[userID].index = idx
+		l.indexes[userID].err = nil
+		l.indexes[userID].updatedAt.Store(startTime.Unix())
+		l.indexesMx.Unlock()
+	}
+
+	// Never return error, otherwise the service terminates.
+	return nil
+}
+
+func (l *Loader) countLoadedIndexesMetric() float64 {
+	l.indexesMx.RLock()
+	defer l.indexesMx.RUnlock()
+
+	count := 0
+	for _, idx := range l.indexes {
+		if idx.index != nil {
+			count++
+		}
+	}
+	return float64(count)
+}
+
+type cachedIndex struct {
+	// We cache either the index or the error occurred while fetching it. They're
+	// mutually exclusive.
+	index *Index
+	err   error
+
+	// Unix timestamp (seconds) of when the index has been updated from the storage the last time.
+	updatedAt *atomic.Int64
```

Comment:
Nit: No need for pointers on atomic values, if outer structure (`cachedIndex`) is used via pointers.